great_expectations/agent/agent.py

```diff
@@ -0,0 +1,201 @@
+        """Schedule _handle_event to run in a thread.
+
+        Callback passed to Subscriber.consume which forwards events to
+        the EventHandler for processing.
+
+        Args:
+            event_context: An Event with related properties and actions.
+        """
+        if self._can_accept_new_task() is not True:
+            # signal to Subscriber that we can't process this message right now
+            return event_context.event_processed(retry=True)
+        self._current_task = self._executor.submit(
+            self._handle_event, event_context=event_context
+        )
+        # TODO lakitu-139: record job as started
+
+        # When the thread exits the results are processed in _handle_event_as_thread_exit.
+        # Curry the event_context, so it's available after the GX process exits.
+        on_exit_callback = partial(
+            self._handle_event_as_thread_exit, event_context=event_context
+        )
+        self._current_task.add_done_callback(on_exit_callback)
+
+    def _handle_event(self, event_context: EventContext) -> EventHandlerResult:
+        """Pass events to EventHandler.
+
+        Callback passed to Subscriber.consume which forwards events to
+        the EventHandler for processing.
+
+        Args:
+            event_context: An Event with related properties and actions.
+        """
+        # warning:  this method will be executed in a different thread than
+        #           where it was defined, so take care with shared resources.
+        #           It's safe to use self._context because we restrict ourselves
+        #           to a single worker thread at any given time. The ack/nack
+        #           callback provided by the Subscriber in event_context will fail,
+        #           since it depends on the channel available in the main thread.
+
+        handler = EventHandler(context=self._context)
+        result = handler.handle_event(event_context=event_context)
+        return result
+
+    def _handle_event_as_thread_exit(
+        self, future: Future, event_context: EventContext
+    ) -> None:
+        """Callback invoked when the thread running GX exits."""
+        # this method will be invoked in the main thread after the worker exits
+
+        # get results or errors from the thread
+        error = future.exception()
+        if error is not None:
+            print("Encountered an error while running job.")
+            print(error)
+        else:
+            print("Job finished successfully.")
+            result = future.result()
+            print(result)
+
+        # TODO lakitu-139: record job as complete and send results
+
+        # ack message and cleanup resources
+        if isinstance(error, UnknownEventError):
+            # We might not have the latest Event definitions, so we requeue
+            # this event in the hope that another agent will understand it.
+            retry = True
```

Comment:
Seems like in this case we should re-queue but also exit the agent and tell the user to update to the latest version. Otherwise the agent will indefinitely loop trying to process with no result

Reply:
my concern with that is if somehow unexpected messages end up passed to the agent it would cause all agents to disconnect, which seems undesirable. my current solution is to allow the same message to be requeued x amount of times (currently set at 10) before sending it to DLX

Reply:
Our current assumption is that the user has precisely 1 agent running. In that scenario, the retries will happen so quickly with the same agent, that the requeueing won't make a difference.

Reply:
sorry wasn't clear in my last message - these retries are set on an async delay, so there's a chance for another agent to pick it up without blocking the thread

Reply:
happy to update this behavior going forward, if we decide this isn't best